src/crypto/hpke/aad.rs

```diff
@@ -0,0 +1,99 @@
+use std::fmt::{Display, Formatter};
+
+use super::{Epoch, KeyIdentifier};
+
+const SALT: &str = "private-attribution";
+
+#[derive(Debug)]
+pub struct HelperOrigin<'a>(pub &'a str);
+#[derive(Debug)]
+pub struct SiteOrigin<'a>(pub &'a str);
+
+/// Represents the [`info`] part of the receiver context, that is: application specific data
+/// for each encryption.
+///
+/// IPA uses key identifier, event epoch and both helper and site origins to authenticate match key
+/// encryptions. It is not guaranteed that the same receiver can be used for anything else.
+///
+/// [`info`]: https://www.rfc-editor.org/rfc/rfc9180.html#name-creating-the-encryption-con
+pub struct AssociatedData<'a> {
+    key_id: KeyIdentifier,
+    epoch: Epoch,
+    helper_origin: &'a HelperOrigin<'a>,
+    site_origin: &'a SiteOrigin<'a>,
+}
+
+#[derive(Debug)]
+pub struct NonAsciiStringError<'a> {
+    input: &'a str,
+}
+
+impl Display for NonAsciiStringError<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "string contains non-ascii symbols: {}", self.input)
+    }
+}
+
+impl<'a> From<&'a str> for NonAsciiStringError<'a> {
+    fn from(input: &'a str) -> Self {
+        Self { input }
+    }
+}
+
+impl<'a> AssociatedData<'a> {
+    /// Creates new instance
+    ///
+    /// ## Errors
+    /// if helper or site origin is not a valid ASCII string.
+    pub fn new(
+        key_id: KeyIdentifier,
+        epoch: Epoch,
+        helper_origin: &'a HelperOrigin<'a>,
+        site_origin: &'a SiteOrigin<'a>,
+    ) -> Result<Self, NonAsciiStringError<'a>> {
+        if !helper_origin.0.is_ascii() {
+            return Err(helper_origin.0.into());
+        }
+
+        if !site_origin.0.is_ascii() {
+            return Err(site_origin.0.into());
+        }
+
+        Ok(Self {
+            key_id,
+            epoch,
+            helper_origin,
+            site_origin,
+        })
+    }
+
+    pub(super) fn key_id(&self) -> KeyIdentifier {
+        self.key_id
+    }
+
+    pub(super) fn epoch(&self) -> Epoch {
+        self.epoch
+    }
+
+    pub(super) fn to_bytes(&self) -> Box<[u8]> {
```

Comment:
```suggestion
    pub(super) fn to_bytes(self) -> Box<[u8]> {
```
Any reason not to consume `self`?

There's a reason I suggest this.  It's possible that we could batch up decryption, but we'd want to have this operation done as few times as possible.  Taking a reference means that you are encouraging people to keep calling this function, when that is just busy work in the end.